--- original/net/minecraft/client/gui/GuiKeyBindingList.java
+++ changed/net/minecraft/client/gui/GuiKeyBindingList.java
@@ -1,6 +1,11 @@
 package net.minecraft.client.gui;
 
 import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import cc.hyperium.Hyperium;
+import cc.hyperium.handlers.handlers.keybinds.HyperiumKeybind;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.resources.I18n;
 import net.minecraft.client.settings.GameSettings;
@@ -21,13 +26,13 @@
         super(mcIn, controls.width, controls.height, 63, controls.height - 32, 20);
         this.field_148191_k = controls;
         this.mc = mcIn;
-        KeyBinding[] akeybinding = (KeyBinding[])ArrayUtils.clone(mcIn.gameSettings.keyBindings);
-        this.listEntries = new GuiListExtended.IGuiListEntry[akeybinding.length + KeyBinding.getKeybinds().size()];
-        Arrays.sort((Object[])akeybinding);
+        List<KeyBinding> allKeys = mc.gameSettings.allKeys;
+        this.listEntries = new GuiListExtended.IGuiListEntry[allKeys.size() + KeyBinding.getKeybinds().size()];
+        Collections.sort(allKeys);
         int i = 0;
         String s = null;
 
-        for (KeyBinding keybinding : akeybinding) {
+        for (KeyBinding keybinding : allKeys) {
             String s1 = keybinding.getKeyCategory();
 
             if (!s1.equals(s)) {
@@ -133,6 +138,13 @@
                         break;
                     }
                 }
+                for (HyperiumKeybind bind : Hyperium.INSTANCE.getHandlers().getKeybindHandler().getKeybinds().values()) {
+                  if (bind.getDescription().equalsIgnoreCase(this.keyDesc)) continue;
+                    if (bind.getKeyCode() == this.keybinding.getKeyCode()) {
+                        flag1 = true;
+                        break;
+                    }
+                }
             }
 
             if (flag) {
@@ -152,6 +164,7 @@
                 GuiKeyBindingList.this.field_148191_k.buttonId = this.keybinding;
                 return true;
             } else if (this.btnReset.mousePressed(GuiKeyBindingList.this.mc, p_148278_2_, p_148278_3_)) {
+                Hyperium.INSTANCE.getHandlers().getKeybindHandler().setBind(this.keyDesc, -1);
                 GuiKeyBindingList.this.mc.gameSettings.setOptionKeyBinding(this.keybinding, this.keybinding.getKeyCodeDefault());
                 KeyBinding.resetKeyBindingArrayAndHash();
                 return true;
