--- original/net/minecraft/client/gui/inventory/GuiInventory.java
+++ changed/net/minecraft/client/gui/inventory/GuiInventory.java
@@ -1,8 +1,12 @@
 package net.minecraft.client.gui.inventory;
 
 import java.io.IOException;
+
+import cc.hyperium.event.EventBus;
+import cc.hyperium.event.interact.ActionPerformedEvent;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.GuiButton;
+import net.minecraft.client.gui.GuiScreen;
 import net.minecraft.client.gui.achievement.GuiAchievements;
 import net.minecraft.client.gui.achievement.GuiStats;
 import net.minecraft.client.gui.inventory.GuiContainerCreative;
@@ -128,6 +132,7 @@
     * Called by the controls from the buttonList when activated. (Mouse pressed for buttons)
     */
    protected void actionPerformed(GuiButton button) throws IOException {
+       EventBus.INSTANCE.post(new ActionPerformedEvent(this, button));
       if(button.id == 0) {
          this.mc.displayGuiScreen(new GuiAchievements(this, this.mc.thePlayer.getStatFileWriter()));
       }
