--- original/net/minecraft/client/settings/GameSettings.java
+++ changed/net/minecraft/client/settings/GameSettings.java
@@ -281,7 +281,7 @@
             this.mc.getTextureMapBlocks().setMipmapLevels(this.mipmapLevels);
             this.mc.getTextureManager().bindTexture(TextureMap.locationBlocksTexture);
             this.mc.getTextureMapBlocks().setBlurMipmapDirect(false, this.mipmapLevels > 0);
-            this.mc.scheduleResourcesRefresh();
+            needsResourceRefresh = true;
          }
       }
 
@@ -1025,6 +1025,15 @@
       return this.useNativeTransport;
    }
 
+   private boolean needsResourceRefresh;
+
+   public void onGuiClosed() {
+      if (needsResourceRefresh) {
+         mc.scheduleResourcesRefresh();
+         needsResourceRefresh = false;
+      }
+   }
+
    @SideOnly(Side.CLIENT)
    public static enum Options {
       INVERT_MOUSE("options.invertMouse", false, true),
