--- original/net/minecraft/client/multiplayer/GuiConnecting.java
+++ changed/net/minecraft/client/multiplayer/GuiConnecting.java
@@ -4,6 +4,9 @@
 import java.net.InetAddress;
 import java.net.UnknownHostException;
 import java.util.concurrent.atomic.AtomicInteger;
+
+import cc.hyperium.event.EventBus;
+import cc.hyperium.event.network.server.ServerJoinEvent;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.GuiButton;
 import net.minecraft.client.gui.GuiDisconnected;
@@ -46,6 +49,7 @@
     }
 
     private void connect(final String ip, final int port) {
+        EventBus.INSTANCE.post(new ServerJoinEvent(ip, port));
         logger.info("Connecting to " + ip + ", " + port);
         (new Thread("Server Connector #" + CONNECTION_ID.incrementAndGet()) {
             public void run() {
