--- original/net/minecraft/client/settings/KeyBinding.java
+++ changed/net/minecraft/client/settings/KeyBinding.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.settings;
 
+import cc.hyperium.Hyperium;
+import cc.hyperium.handlers.handlers.keybinds.HyperiumKeybind;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import java.util.List;
@@ -8,6 +10,7 @@
 import net.minecraft.util.IntHashMap;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.lwjgl.input.Keyboard;
 
 @SideOnly(Side.CLIENT)
 public class KeyBinding implements Comparable<KeyBinding> {
@@ -60,6 +63,11 @@
         return keybindSet;
     }
 
+    public KeyBinding(String description, int keyCode, String category, int defaultKeyCode) {
+        this(description, defaultKeyCode, category);
+        this.keyCode = keyCode;
+    }
+
     public KeyBinding(String description, int keyCode, String category) {
         this.keyDescription = description;
         this.keyCode = keyCode;
@@ -123,4 +131,18 @@
 
         return i;
     }
+
+    public void setPressed(boolean pressed) {
+        this.pressed = pressed;
+    }
+
+    // Hyperium: Add updateKeybindState - this is used to fix keybinds not registering when closing a GUI
+    public static void updateKeyBindState() {
+        for (KeyBinding keybinding : keybindArray) {
+            try {
+                setKeyBindState(keybinding.getKeyCode(), keybinding.getKeyCode() < 256 && Keyboard.isKeyDown(keybinding.getKeyCode()));
+            } catch (IndexOutOfBoundsException var3) {
+            }
+        }
+    }
 }
