--- original/net/minecraft/world/WorldProvider.java
+++ changed/net/minecraft/world/WorldProvider.java
@@ -1,5 +1,6 @@
 package net.minecraft.world;
 
+import cc.hyperium.handlers.handlers.cloud.IRenderHandler;
 import net.minecraft.init.Blocks;
 import net.minecraft.util.BlockPos;
 import net.minecraft.util.MathHelper;
@@ -245,4 +246,14 @@
    public WorldBorder getWorldBorder() {
       return new WorldBorder();
    }
+
+   private IRenderHandler cloudRenderer;
+
+   public IRenderHandler getCloudRenderer() {
+      return cloudRenderer;
+   }
+
+   public void setCloudRenderer(IRenderHandler cloudRenderer) {
+      this.cloudRenderer = cloudRenderer;
+   }
 }
