--- original/net/minecraft/crash/CrashReport.java
+++ changed/net/minecraft/crash/CrashReport.java
@@ -1,5 +1,6 @@
 package net.minecraft.crash;
 
+import cc.hyperium.hooks.CrashReportHook;
 import com.google.common.collect.Lists;
 import java.io.File;
 import java.io.FileWriter;
@@ -28,7 +29,7 @@
     /** The Throwable that is the "cause" for this crash and Crash Report. */
     private final Throwable cause;
     /** Category of crash */
-    private final CrashReportCategory theReportCategory = new CrashReportCategory(this, "System Details");
+    public final CrashReportCategory theReportCategory = new CrashReportCategory(this, "System Details");
     private final List<CrashReportCategory> crashReportSections = Lists.<CrashReportCategory>newArrayList();
     /** File of crash report. */
     private File crashReportFile;
@@ -46,6 +47,7 @@
      * Populates this crash report with initial information about the running server and operating system / java environment
      */
     private void populateEnvironment() {
+        CrashReportHook.populateEnvironment(this);
         this.theReportCategory.addCrashSectionCallable("Minecraft Version", new Callable<String>() {
             public String call() {
                 return "1.8.9";
