--- original/net/minecraft/client/network/NetworkPlayerInfo.java
+++ changed/net/minecraft/client/network/NetworkPlayerInfo.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.network;
 
+import cc.hyperium.mods.nickhider.NickHider;
+import cc.hyperium.mods.nickhider.config.NickHiderConfig;
 import com.google.common.base.Objects;
 import com.mojang.authlib.GameProfile;
 import com.mojang.authlib.minecraft.MinecraftProfileTexture;
@@ -22,9 +24,9 @@
     private WorldSettings.GameType gameType;
     /** Player response time to server in milliseconds */
     private int responseTime;
-    private boolean playerTexturesLoaded = false;
-    private ResourceLocation locationSkin;
-    private ResourceLocation locationCape;
+    public boolean playerTexturesLoaded = false;
+    public ResourceLocation locationSkin;
+    public ResourceLocation locationCape;
     private String skinType;
     /** When this is non-null, it is displayed instead of the player's real name */
     private IChatComponent displayName;
@@ -73,10 +75,38 @@
     }
 
     public String getSkinType() {
+        NickHider instance = NickHider.instance;
+
+        if (instance != null && instance.getNickHiderConfig().isHideSkins() && instance.getNickHiderConfig().isMasterEnabled()) {
+            NickHiderConfig config = instance.getNickHiderConfig();
+
+            if (gameProfile.getId().equals(Minecraft.getMinecraft().thePlayer.getUniqueID())) {
+                if (config.isUseRealSkinForSelf() && instance.getPlayerSkin() != null) {
+                    return instance.getPlayerRealSkinType();
+                }
+            } else if (config.isHideOtherSkins()) {
+                if (config.isUsePlayerSkinForAll() && instance.getPlayerSkin() != null) {
+                    return instance.getPlayerRealSkinType();
+                }
+            }
+        }
+
         return this.skinType == null ? DefaultPlayerSkin.getSkinType(this.gameProfile.getId()) : this.skinType;
     }
 
     public ResourceLocation getLocationSkin() {
+        NickHider instance = NickHider.instance;
+
+        if (instance != null && instance.getNickHiderConfig().isHideSkins() && instance.getNickHiderConfig().isMasterEnabled()) {
+            NickHiderConfig config = instance.getNickHiderConfig();
+
+            if (gameProfile.getId().equals(Minecraft.getMinecraft().thePlayer.getUniqueID())) {
+                return config.isUseRealSkinForSelf() && instance.getPlayerSkin() != null ? instance.getPlayerSkin() : DefaultPlayerSkin.getDefaultSkin(gameProfile.getId());
+            } else if (config.isHideOtherSkins()) {
+                return config.isUsePlayerSkinForAll() && instance.getPlayerSkin() != null ? instance.getPlayerSkin() : DefaultPlayerSkin.getDefaultSkin(gameProfile.getId());
+            }
+        }
+
         if (this.locationSkin == null) {
             this.loadPlayerTextures();
         }
@@ -85,6 +115,16 @@
     }
 
     public ResourceLocation getLocationCape() {
+        NickHider instance = NickHider.instance;
+
+        if (instance != null && instance.getNickHiderConfig().isHideSkins() && instance.getNickHiderConfig().isMasterEnabled()) {
+            NickHiderConfig config = instance.getNickHiderConfig();
+
+            if (gameProfile.getId().equals(Minecraft.getMinecraft().thePlayer.getUniqueID()) && config.isUseRealSkinForSelf()) {
+                return instance.getPlayerCape();
+            }
+        }
+
         if (this.locationCape == null) {
             this.loadPlayerTextures();
         }
