--- original/net/minecraft/inventory/ContainerEnchantment.java
+++ changed/net/minecraft/inventory/ContainerEnchantment.java
@@ -327,9 +327,8 @@
                return null;
             }
 
-            if(itemstack1.hasTagCompound() && itemstack1.stackSize == 1) {
-               ((Slot)this.inventorySlots.get(0)).putStack(itemstack1.copy());
-               itemstack1.stackSize = 0;
+            if(itemstack1.hasTagCompound()) {
+               this.inventorySlots.get(0).putStack(itemstack1.splitStack(1));
             } else if(itemstack1.stackSize >= 1) {
                ((Slot)this.inventorySlots.get(0)).putStack(new ItemStack(itemstack1.getItem(), 1, itemstack1.getMetadata()));
                --itemstack1.stackSize;
