--- original/net/minecraft/client/gui/GuiControls.java
+++ changed/net/minecraft/client/gui/GuiControls.java
@@ -1,6 +1,10 @@
 package net.minecraft.client.gui;
 
 import java.io.IOException;
+
+import cc.hyperium.Hyperium;
+import cc.hyperium.gui.keybinds.GuiKeybinds;
+import cc.hyperium.handlers.handlers.keybinds.HyperiumKeybind;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.resources.I18n;
 import net.minecraft.client.settings.GameSettings;
@@ -67,7 +71,11 @@
             for (KeyBinding keybinding : this.mc.gameSettings.keyBindings) {
                 keybinding.setKeyCode(keybinding.getKeyCodeDefault());
             }
+            for (HyperiumKeybind bind : Hyperium.INSTANCE.getHandlers().getKeybindHandler().getKeybinds().values()) {
+                bind.setKeyCode(bind.getDefaultKeyCode());
+            }
 
+            Hyperium.INSTANCE.getHandlers().getKeybindHandler().refreshKeys();
             KeyBinding.resetKeyBindingArrayAndHash();
         } else if (button.id < 100 && button instanceof GuiOptionButton) {
             this.options.setOptionValue(((GuiOptionButton)button).returnEnumOptions(), 1);
@@ -106,6 +114,7 @@
             if (keyCode == 1) {
                 this.options.setOptionKeyBinding(this.buttonId, 0);
             } else if (keyCode != 0) {
+                Hyperium.INSTANCE.getHandlers().getKeybindHandler().setBind(this.buttonId.getKeyDescription(), keyCode);
                 this.options.setOptionKeyBinding(this.buttonId, keyCode);
             } else if (typedChar > 0) {
                 this.options.setOptionKeyBinding(this.buttonId, typedChar + 256);
@@ -119,6 +128,11 @@
         }
     }
 
+    @Override
+    public void onGuiClosed() {
+        Hyperium.INSTANCE.getHandlers().getKeybindHandler().getConfig().save();
+    }
+
     /**
      * Draws the screen and all the components in it. Args : mouseX, mouseY, renderPartialTicks
      */
@@ -135,6 +149,13 @@
             }
         }
 
+        for (HyperiumKeybind bind : Hyperium.INSTANCE.getHandlers().getKeybindHandler().getKeybinds().values()) {
+            if (bind.getKeyCode() != bind.getDefaultKeyCode()) {
+                flag = false;
+                break;
+            }
+        }
+
         this.buttonReset.enabled = !flag;
         super.drawScreen(mouseX, mouseY, partialTicks);
     }
