--- original/net/minecraft/client/renderer/entity/RenderManager.java
+++ changed/net/minecraft/client/renderer/entity/RenderManager.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.renderer.entity;
 
+import cc.hyperium.cosmetics.companions.hamster.EntityHamster;
+import cc.hyperium.cosmetics.companions.hamster.RenderHamster;
 import com.google.common.collect.Maps;
 import java.util.Map;
 import net.minecraft.block.Block;
@@ -249,6 +251,8 @@
       this.playerRenderer = new RenderPlayer(this);
       this.skinMap.put("default", this.playerRenderer);
       this.skinMap.put("slim", new RenderPlayer(this, true));
+
+      entityRenderMap.put(EntityHamster.class, new RenderHamster(this));
    }
 
    public void setRenderPosition(double renderPosXIn, double renderPosYIn, double renderPosZIn) {
@@ -488,4 +492,20 @@
    public void setRenderOutlines(boolean renderOutlinesIn) {
       this.renderOutlines = renderOutlinesIn;
    }
+
+   public double getPosX() {
+      return renderPosX;
+   }
+
+   public double getPosY() {
+      return renderPosY;
+   }
+
+   public double getPosZ() {
+      return renderPosZ;
+   }
+
+   public Map<String, RenderPlayer> getSkinMap() {
+      return skinMap;
+   }
 }
