--- original/net/minecraft/nbt/NBTUtil.java
+++ changed/net/minecraft/nbt/NBTUtil.java
@@ -24,11 +24,13 @@
         if (StringUtils.isNullOrEmpty(s) && StringUtils.isNullOrEmpty(s1)) {
             return null;
         } else {
-            UUID uuid;
-            try {
-                uuid = UUID.fromString(s1);
-            } catch (Throwable var12) {
-                uuid = null;
+            UUID uuid = null;
+
+            if (s1 != null) {
+                try {
+                    uuid = UUID.fromString(s1);
+                } catch (Throwable ignored) {
+                }
             }
 
             GameProfile gameprofile = new GameProfile(uuid, s);
