--- original/net/minecraft/server/MinecraftServer.java
+++ changed/net/minecraft/server/MinecraftServer.java
@@ -554,6 +554,7 @@
                 ImageIO.write(bufferedimage, "PNG", (OutputStream)(new ByteBufOutputStream(bytebuf)));
                 ByteBuf bytebuf1 = Base64.encode(bytebuf);
                 response.setFavicon("data:image/png;base64," + bytebuf1.toString(Charsets.UTF_8));
+                bytebuf1.release();
             } catch (Exception exception) {
                 logger.error((String)"Couldn\'t load server icon", (Throwable)exception);
             } finally {
