--- original/net/minecraft/entity/EntityLivingBase.java
+++ changed/net/minecraft/entity/EntityLivingBase.java
@@ -1,5 +1,7 @@
 package net.minecraft.entity;
 
+import cc.hyperium.event.EventBus;
+import cc.hyperium.event.entity.LivingDeathEvent;
 import com.google.common.base.Predicate;
 import com.google.common.base.Predicates;
 import com.google.common.collect.Maps;
@@ -12,6 +14,7 @@
 import net.minecraft.block.Block;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
+import net.minecraft.client.entity.EntityPlayerSP;
 import net.minecraft.enchantment.EnchantmentHelper;
 import net.minecraft.entity.ai.attributes.AttributeModifier;
 import net.minecraft.entity.ai.attributes.BaseAttributeMap;
@@ -817,6 +820,8 @@
      * Called when the mob's health reaches 0.
      */
     public void onDeath(DamageSource cause) {
+        EventBus.INSTANCE.post(new LivingDeathEvent(this, cause));
+
         Entity entity = cause.getEntity();
         EntityLivingBase entitylivingbase = this.getAttackingEntity();
 
@@ -1783,6 +1788,10 @@
      * interpolated look vector
      */
     public Vec3 getLook(float partialTicks) {
+        if (this instanceof EntityPlayerSP) {
+            return super.getLook(partialTicks);
+        }
+
         if (partialTicks == 1.0F) {
             return this.getVectorForRotation(this.rotationPitch, this.rotationYawHead);
         } else {
