--- original/net/minecraft/client/gui/GuiIngame.java
+++ changed/net/minecraft/client/gui/GuiIngame.java
@@ -6,6 +6,12 @@
 import java.util.Collection;
 import java.util.List;
 import java.util.Random;
+import cc.hyperium.Hyperium;
+import cc.hyperium.config.Settings;
+import cc.hyperium.event.EventBus;
+import cc.hyperium.event.render.RenderHUDEvent;
+import cc.hyperium.event.render.RenderSelectedItemEvent;
+import cc.hyperium.mods.chromahud.displayitems.hyperium.ScoreboardDisplay;
 import net.minecraft.block.material.Material;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.GlStateManager;
@@ -90,6 +96,11 @@
     /** Used with updateCounter to make the heart bar flash */
     private long healthUpdateCounter = 0L;
 
+    public static boolean renderScoreboard = true;
+    public static boolean renderHealth = true;
+    public static boolean renderFood = true;
+    public static boolean renderArmor = true;
+
     public GuiIngame(Minecraft mcIn) {
         this.mc = mcIn;
         this.itemRenderer = mcIn.getRenderItem();
@@ -299,6 +310,11 @@
         GlStateManager.popMatrix();
         scoreobjective1 = scoreboard.getObjectiveInDisplaySlot(0);
 
+        mc.mcProfiler.startSection("hyperium_overlay");
+        EventBus.INSTANCE.post(new RenderHUDEvent(new ScaledResolution(mc), partialTicks));
+        GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+        mc.mcProfiler.endSection();
+
         if (!this.mc.gameSettings.keyBindPlayerList.isKeyDown() || this.mc.isIntegratedServerRunning() && this.mc.thePlayer.sendQueue.getPlayerInfoMap().size() <= 1 && scoreobjective1 == null) {
             this.overlayPlayerList.updatePlayerList(false);
         } else {
@@ -418,6 +434,10 @@
                 GlStateManager.tryBlendFuncSeparate(770, 771, 1, 0);
                 this.getFontRenderer().drawStringWithShadow(s, (float)i, (float)j, 16777215 + (k << 24));
                 GlStateManager.disableBlend();
+
+                EventBus.INSTANCE.post(new RenderSelectedItemEvent(scaledRes));
+                GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+
                 GlStateManager.popMatrix();
             }
         }
@@ -441,6 +461,9 @@
     }
 
     protected boolean showCrosshair() {
+        if (mc.gameSettings.thirdPersonView > 0 && Settings.CROSSHAIR_PERSPECTIVE) {
+            return false;
+        }
         if (this.mc.gameSettings.showDebugInfo && !this.mc.thePlayer.hasReducedDebug() && !this.mc.gameSettings.reducedDebugInfo) {
             return false;
         } else if (this.mc.playerController.isSpectator()) {
@@ -467,52 +490,13 @@
     }
 
     private void renderScoreboard(ScoreObjective objective, ScaledResolution scaledRes) {
-        Scoreboard scoreboard = objective.getScoreboard();
-        Collection<Score> collection = scoreboard.getSortedScores(objective);
-        List<Score> list = Lists.newArrayList(Iterables.filter(collection, new Predicate<Score>() {
-            public boolean apply(Score p_apply_1_) {
-                return p_apply_1_.getPlayerName() != null && !p_apply_1_.getPlayerName().startsWith("#");
-            }
-        }));
-
-        if (list.size() > 15) {
-            collection = Lists.newArrayList(Iterables.skip(list, collection.size() - 15));
-        } else {
-            collection = list;
+        if (!renderScoreboard) {
+            return;
         }
 
-        int i = this.getFontRenderer().getStringWidth(objective.getDisplayName());
-
-        for (Score score : collection) {
-            ScorePlayerTeam scoreplayerteam = scoreboard.getPlayersTeam(score.getPlayerName());
-            String s = ScorePlayerTeam.formatPlayerName(scoreplayerteam, score.getPlayerName()) + ": " + EnumChatFormatting.RED + score.getScorePoints();
-            i = Math.max(i, this.getFontRenderer().getStringWidth(s));
-        }
-
-        int i1 = collection.size() * this.getFontRenderer().FONT_HEIGHT;
-        int j1 = scaledRes.getScaledHeight() / 2 + i1 / 3;
-        int k1 = 3;
-        int l1 = scaledRes.getScaledWidth() - i - k1;
-        int j = 0;
-
-        for (Score score1 : collection) {
-            ++j;
-            ScorePlayerTeam scoreplayerteam1 = scoreboard.getPlayersTeam(score1.getPlayerName());
-            String s1 = ScorePlayerTeam.formatPlayerName(scoreplayerteam1, score1.getPlayerName());
-            String s2 = EnumChatFormatting.RED + "" + score1.getScorePoints();
-            int k = j1 - j * this.getFontRenderer().FONT_HEIGHT;
-            int l = scaledRes.getScaledWidth() - k1 + 2;
-            drawRect(l1 - 2, k, l, k + this.getFontRenderer().FONT_HEIGHT, 1342177280);
-            this.getFontRenderer().drawString(s1, l1, k, 553648127);
-            this.getFontRenderer().drawString(s2, l - this.getFontRenderer().getStringWidth(s2), k, 553648127);
-
-            if (j == collection.size()) {
-                String s3 = objective.getDisplayName();
-                drawRect(l1 - 2, k - this.getFontRenderer().FONT_HEIGHT - 1, l, k - 1, 1610612736);
-                drawRect(l1 - 2, k - 1, l, k, 1342177280);
-                this.getFontRenderer().drawString(s3, l1 + i / 2 - this.getFontRenderer().getStringWidth(s3) / 2, k - this.getFontRenderer().FONT_HEIGHT, 553648127);
-            }
-        }
+        ScoreboardDisplay.objective = objective;
+        ScoreboardDisplay.resolution = scaledRes;
+        Hyperium.INSTANCE.getHandlers().getScoreboardRenderer().render(objective, scaledRes);
     }
 
     private void renderPlayerStats(ScaledResolution scaledRes) {
@@ -561,6 +545,7 @@
 
             this.mc.mcProfiler.startSection("armor");
 
+            if (renderArmor) {
             for (int i3 = 0; i3 < 10; ++i3) {
                 if (k2 > 0) {
                     int j3 = i1 + i3 * 8;
@@ -578,6 +563,7 @@
                     }
                 }
             }
+            } // renderArmor end
 
             this.mc.mcProfiler.endStartSection("health");
 
@@ -614,8 +600,10 @@
                     k4 = 5;
                 }
 
+                if (renderHealth) {
                 this.drawTexturedModalRect(i4, j4, 16 + k3 * 9, 9 * k4, 9, 9);
 
+                if (!Settings.OLD_HEALTH) {
                 if (flag) {
                     if (i6 * 2 + 1 < j) {
                         this.drawTexturedModalRect(i4, j4, j6 + 54, 9 * k4, 9, 9);
@@ -625,6 +613,7 @@
                         this.drawTexturedModalRect(i4, j4, j6 + 63, 9 * k4, 9, 9);
                     }
                 }
+                } // old health end
 
                 if (f2 > 0.0F) {
                     if (f2 == f1 && f1 % 2.0F == 1.0F) {
@@ -643,8 +632,10 @@
                         this.drawTexturedModalRect(i4, j4, j6 + 45, 9 * k4, 9, 9);
                     }
                 }
+                } // renderHealth end
             }
 
+            if (renderFood) {
             Entity entity = entityplayer.ridingEntity;
 
             if (entity == null) {
@@ -729,6 +720,7 @@
                     j9 -= 10;
                 }
             }
+            } // renderFood end
 
             this.mc.mcProfiler.endStartSection("air");
 
@@ -754,26 +746,43 @@
      * Renders dragon's (boss) health on the HUD
      */
     private void renderBossHealth() {
+        if (!Settings.BOSSBAR_ALL) return;
+
         if (BossStatus.bossName != null && BossStatus.statusBarTime > 0) {
             --BossStatus.statusBarTime;
+
             FontRenderer fontrenderer = this.mc.fontRendererObj;
             ScaledResolution scaledresolution = new ScaledResolution(this.mc);
-            int i = scaledresolution.getScaledWidth();
-            int j = 182;
-            int k = i / 2 - j / 2;
-            int l = (int)(BossStatus.healthScale * (float)(j + 1));
-            int i1 = 12;
-            this.drawTexturedModalRect(k, i1, 0, 74, j, 5);
-            this.drawTexturedModalRect(k, i1, 0, 74, j, 5);
+            double scaledWidth = scaledresolution.getScaledWidth();
+            double scaledHeight = scaledresolution.getScaledHeight();
 
-            if (l > 0) {
-                this.drawTexturedModalRect(k, i1, 0, 79, l, 5);
+            String bossName = BossStatus.bossName;
+
+            if (Settings.BOSSBAR_TEXT) {
+                GlStateManager.pushMatrix();
+                GlStateManager.translate(scaledWidth * Settings.BOSSBAR_X, scaledHeight * Settings.BOSSBAR_Y - (10D * Settings.BOSSBAR_SCALE), 0);
+                GlStateManager.scale(Settings.BOSSBAR_SCALE, Settings.BOSSBAR_SCALE, Settings.BOSSBAR_SCALE);
+                GlStateManager.translate(-fontrenderer.getStringWidth(bossName) / 2F , 0, 0);
+                fontrenderer.drawStringWithShadow(bossName, 0, 0, -1);
+                GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+                Minecraft.getMinecraft().getTextureManager().bindTexture(Gui.icons);
+                GlStateManager.popMatrix();
             }
 
-            String s = BossStatus.bossName;
-            this.getFontRenderer().drawStringWithShadow(s, (float)(i / 2 - this.getFontRenderer().getStringWidth(s) / 2), (float)(i1 - 10), 16777215);
-            GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
-            this.mc.getTextureManager().bindTexture(icons);
+            int widthLocation = 182;
+
+            if (Settings.BOSSBAR_BAR) {
+                int healthScale = (int) (BossStatus.healthScale * (float) (widthLocation + 1) * Settings.BOSSBAR_SCALE);
+
+                GlStateManager.pushMatrix();
+                GlStateManager.translate(Settings.BOSSBAR_X * scaledWidth - widthLocation / 2F * Settings.BOSSBAR_SCALE, Settings.BOSSBAR_Y * scaledHeight , 0);
+                GlStateManager.scale(Settings.BOSSBAR_SCALE, Settings.BOSSBAR_SCALE, Settings.BOSSBAR_SCALE);
+
+                drawTexturedModalRect(0, 0, 0, 74, widthLocation, 5);
+
+                if (healthScale > 0) drawTexturedModalRect(0, 0, 0, 79, widthLocation, 5);
+                GlStateManager.popMatrix();
+            }
         }
     }
 
@@ -948,6 +957,8 @@
     }
 
     public void displayTitle(String title, String subTitle, int timeFadeIn, int displayTime, int timeFadeOut) {
+        if (Settings.HIDE_TITLES) return;
+
         if (title == null && subTitle == null && timeFadeIn < 0 && displayTime < 0 && timeFadeOut < 0) {
             this.displayedTitle = "";
             this.displayedSubTitle = "";
