--- original/net/minecraft/world/gen/feature/WorldGenBigTree.java
+++ changed/net/minecraft/world/gen/feature/WorldGenBigTree.java
@@ -256,12 +256,14 @@
       }
 
       if(!this.validTreeLocation()) {
+         world = null;
          return false;
       } else {
          this.generateLeafNodeList();
          this.generateLeaves();
          this.generateTrunk();
          this.generateLeafNodeBases();
+         world = null;
          return true;
       }
    }
