--- original/net/minecraft/client/gui/GuiScreenResourcePacks.java
+++ changed/net/minecraft/client/gui/GuiScreenResourcePacks.java
@@ -1,5 +1,6 @@
 package net.minecraft.client.gui;
 
+import cc.hyperium.gui.resource.ResourcePackUtil;
 import com.google.common.collect.Lists;
 import java.io.File;
 import java.io.IOException;
@@ -30,6 +31,8 @@
     private GuiResourcePackSelected selectedResourcePacksList;
     private boolean changed = false;
 
+    private GuiTextField textField;
+
     public GuiScreenResourcePacks(GuiScreen parentScreenIn) {
         this.parentScreen = parentScreenIn;
     }
@@ -39,6 +42,10 @@
      * resizes, the buttonList is cleared beforehand.
      */
     public void initGui() {
+        String s1 = textField == null ? "" : textField.getText();
+        textField = new GuiTextField(3, fontRendererObj, width / 2 - 4 - 200, height - 24, 200, 20);
+        textField.setText(s1);
+
         this.buttonList.add(new GuiOptionButton(2, this.width / 2 - 154, this.height - 48, I18n.format("resourcePack.openFolder", new Object[0])));
         this.buttonList.add(new GuiOptionButton(1, this.width / 2 + 4, this.height - 48, I18n.format("gui.done", new Object[0])));
 
@@ -67,6 +74,14 @@
         this.selectedResourcePacksList = new GuiResourcePackSelected(this.mc, 200, this.height, this.selectedResourcePacks);
         this.selectedResourcePacksList.setSlotXBoundsFromLeft(this.width / 2 + 4);
         this.selectedResourcePacksList.registerScrollButtons(7, 8);
+
+        for (GuiButton b : buttonList) {
+            b.setWidth(200);
+
+            if (b.id == 2) {
+                b.xPosition = width / 2 - 204;
+            }
+        }
     }
 
     /**
@@ -176,8 +191,17 @@
         super.mouseClicked(mouseX, mouseY, mouseButton);
         this.availableResourcePacksList.mouseClicked(mouseX, mouseY, mouseButton);
         this.selectedResourcePacksList.mouseClicked(mouseX, mouseY, mouseButton);
+        if (textField != null) textField.mouseClicked(mouseX, mouseY, mouseButton);
     }
 
+    @Override
+    protected void keyTyped(char typedChar, int keyCode) throws IOException {
+        super.keyTyped(typedChar, keyCode);
+
+        if (textField.isFocused()) textField.textboxKeyTyped(typedChar, keyCode);
+        availableResourcePacksList = ResourcePackUtil.updateList(textField, availableResourcePacksList, availableResourcePacks, mc, width, height);
+    }
+
     /**
      * Called when a mouse button is released.  Args : mouseX, mouseY, releaseButton
      */
@@ -193,8 +217,8 @@
         this.availableResourcePacksList.drawScreen(mouseX, mouseY, partialTicks);
         this.selectedResourcePacksList.drawScreen(mouseX, mouseY, partialTicks);
         this.drawCenteredString(this.fontRendererObj, I18n.format("resourcePack.title", new Object[0]), this.width / 2, 16, 16777215);
-        this.drawCenteredString(this.fontRendererObj, I18n.format("resourcePack.folderInfo", new Object[0]), this.width / 2 - 77, this.height - 26, 8421504);
         super.drawScreen(mouseX, mouseY, partialTicks);
+        textField.drawTextBox();
     }
 
     /**
