--- original/net/minecraft/item/ItemBow.java
+++ changed/net/minecraft/item/ItemBow.java
@@ -1,5 +1,7 @@
 package net.minecraft.item;
 
+import cc.hyperium.event.EventBus;
+import cc.hyperium.event.entity.ArrowShootEvent;
 import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.enchantment.Enchantment;
 import net.minecraft.enchantment.EnchantmentHelper;
@@ -73,6 +75,7 @@
             if (!worldIn.isRemote) {
                 worldIn.spawnEntityInWorld(entityarrow);
             }
+            EventBus.INSTANCE.post(new ArrowShootEvent(entityarrow, i, stack));
         }
     }
 
