--- original/net/minecraft/client/renderer/entity/RenderManager.java
+++ changed/net/minecraft/client/renderer/entity/RenderManager.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.renderer.entity;
 
+import cc.hyperium.cosmetics.companions.hamster.EntityHamster;
+import cc.hyperium.cosmetics.companions.hamster.RenderHamster;
 import com.google.common.collect.Maps;
 import java.util.Map;
 import net.minecraft.block.Block;
@@ -106,13 +108,13 @@
 @SideOnly(Side.CLIENT)
 public class RenderManager {
     private Map < Class <? extends Entity > , Render <? extends Entity >> entityRenderMap = Maps. < Class <? extends Entity > , Render <? extends Entity >> newHashMap();
-    private Map<String, RenderPlayer> skinMap = Maps.<String, RenderPlayer>newHashMap();
+    public Map<String, RenderPlayer> skinMap = Maps.<String, RenderPlayer>newHashMap();
     private RenderPlayer playerRenderer;
     /** Renders fonts */
     private FontRenderer textRenderer;
-    private double renderPosX;
-    private double renderPosY;
-    private double renderPosZ;
+    public double renderPosX;
+    public double renderPosY;
+    public double renderPosZ;
     public TextureManager renderEngine;
     /** Reference to the World object. */
     public World worldObj;
@@ -195,6 +197,8 @@
         this.playerRenderer = new RenderPlayer(this);
         this.skinMap.put("default", this.playerRenderer);
         this.skinMap.put("slim", new RenderPlayer(this, true));
+
+        entityRenderMap.put(EntityHamster.class, new RenderHamster(this));
     }
 
     public void setRenderPosition(double renderPosXIn, double renderPosYIn, double renderPosZIn) {
