--- original/net/minecraft/nbt/NBTUtil.java
+++ changed/net/minecraft/nbt/NBTUtil.java
@@ -15,39 +15,34 @@
    public static GameProfile readGameProfileFromNBT(NBTTagCompound compound) {
       String s = null;
       String s1 = null;
-      if(compound.hasKey("Name", 8)) {
-         s = compound.getString("Name");
-      }
 
-      if(compound.hasKey("Id", 8)) {
-         s1 = compound.getString("Id");
-      }
+      if (compound.hasKey("Name", 8)) s = compound.getString("Name");
+      if (compound.hasKey("Id", 8)) s1 = compound.getString("Id");
 
-      if(StringUtils.isNullOrEmpty(s) && StringUtils.isNullOrEmpty(s1)) {
+      if (StringUtils.isNullOrEmpty(s) && StringUtils.isNullOrEmpty(s1)) {
          return null;
       } else {
-         UUID uuid;
-         try {
-            uuid = UUID.fromString(s1);
-         } catch (Throwable var12) {
-            uuid = null;
-         }
+         UUID uuid = null;
+         if (s1 != null)
+            try {
+               uuid = UUID.fromString(s1);
+            } catch (Throwable ignored) {
+            }
 
          GameProfile gameprofile = new GameProfile(uuid, s);
-         if(compound.hasKey("Properties", 10)) {
+
+         if (compound.hasKey("Properties", 10)) {
             NBTTagCompound nbttagcompound = compound.getCompoundTag("Properties");
 
-            for(String s2 : nbttagcompound.getKeySet()) {
+            for (String s2 : nbttagcompound.getKeySet()) {
                NBTTagList nbttaglist = nbttagcompound.getTagList(s2, 10);
 
-               for(int i = 0; i < nbttaglist.tagCount(); ++i) {
+               int bound = nbttaglist.tagCount();
+               for (int i = 0; i < bound; i++) {
                   NBTTagCompound nbttagcompound1 = nbttaglist.getCompoundTagAt(i);
                   String s3 = nbttagcompound1.getString("Value");
-                  if(nbttagcompound1.hasKey("Signature", 8)) {
-                     gameprofile.getProperties().put(s2, new Property(s2, s3, nbttagcompound1.getString("Signature")));
-                  } else {
-                     gameprofile.getProperties().put(s2, new Property(s2, s3));
-                  }
+                  gameprofile.getProperties().put(s2, nbttagcompound1.hasKey("Signature", 8) ?
+                          new Property(s2, s3, nbttagcompound1.getString("Signature")) : new Property(s2, s3));
                }
             }
          }
