--- original/net/minecraft/client/renderer/ChunkRenderContainer.java
+++ changed/net/minecraft/client/renderer/ChunkRenderContainer.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.renderer;
 
+import cc.hyperium.Hyperium;
+import cc.hyperium.mods.chunkanimator.AnimationHandler;
 import com.google.common.collect.Lists;
 import java.util.List;
 import net.minecraft.client.renderer.GlStateManager;
@@ -17,6 +19,8 @@
    protected List<RenderChunk> renderChunks = Lists.<RenderChunk>newArrayListWithCapacity(17424);
    protected boolean initialized;
 
+   private AnimationHandler handler;
+
    public void initialize(double viewEntityXIn, double viewEntityYIn, double viewEntityZIn) {
       this.initialized = true;
       this.renderChunks.clear();
@@ -26,6 +30,8 @@
    }
 
    public void preRenderChunk(RenderChunk renderChunkIn) {
+      if (handler == null) handler = Hyperium.INSTANCE.getModIntegration().getChunkAnimator().getAnimationHandler();
+      handler.preRenderChunk(renderChunkIn);
       BlockPos blockpos = renderChunkIn.getPosition();
       GlStateManager.translate((float)((double)blockpos.getX() - this.viewEntityX), (float)((double)blockpos.getY() - this.viewEntityY), (float)((double)blockpos.getZ() - this.viewEntityZ));
    }
