--- original/net/minecraft/client/gui/GuiOptions.java
+++ changed/net/minecraft/client/gui/GuiOptions.java
@@ -1,6 +1,8 @@
 package net.minecraft.client.gui;
 
 import java.io.IOException;
+
+import cc.hyperium.gui.hyperium.HyperiumMainGui;
 import net.minecraft.client.audio.PositionedSoundRecord;
 import net.minecraft.client.audio.SoundCategory;
 import net.minecraft.client.audio.SoundEventAccessorComposite;
@@ -104,6 +106,15 @@
       this.buttonList.add(new GuiButton(105, this.width / 2 - 155, this.height / 6 + 144 - 6, 150, 20, I18n.format("options.resourcepack", new Object[0])));
       this.buttonList.add(new GuiButton(104, this.width / 2 + 5, this.height / 6 + 144 - 6, 150, 20, I18n.format("options.snooper.view", new Object[0])));
       this.buttonList.add(new GuiButton(200, this.width / 2 - 100, this.height / 6 + 168, I18n.format("gui.done", new Object[0])));
+
+      for (GuiButton b : buttonList) {
+         if (b.id == 200) {
+            b.yPosition = height - 30;
+         }
+      }
+
+      buttonList.add(new GuiButton(114514, width / 2 - 155, height / 6 + 18, 150, 20,
+              I18n.format("button.ingame.hyperiumsettings")));
    }
 
    public String func_175355_a(EnumDifficulty p_175355_1_) {
@@ -204,7 +215,9 @@
             }
          }
 
+         if (button.id == 114514) HyperiumMainGui.INSTANCE.show();
       }
+
    }
 
    /**
