--- original/net/minecraft/client/resources/Locale.java
+++ changed/net/minecraft/client/resources/Locale.java
@@ -1,21 +1,22 @@
 package net.minecraft.client.resources;
 
+import cc.hyperium.utils.locale.HyperiumLocale;
 import com.google.common.base.Splitter;
 import com.google.common.collect.Iterables;
 import com.google.common.collect.Maps;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
+import org.apache.commons.io.Charsets;
+import org.apache.commons.io.IOUtils;
+
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.IllegalFormatException;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Supplier;
 import java.util.regex.Pattern;
-import net.minecraft.client.resources.IResource;
-import net.minecraft.client.resources.IResourceManager;
-import net.minecraft.util.ResourceLocation;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
-import org.apache.commons.io.Charsets;
-import org.apache.commons.io.IOUtils;
 
 @SideOnly(Side.CLIENT)
 public class Locale {
@@ -34,6 +35,8 @@
       for(String s : languageList) {
          String s1 = String.format("lang/%s.lang", new Object[]{s});
 
+         injectI18nData(s1, s);
+
          for(String s2 : resourceManager.getResourceDomains()) {
             try {
                this.loadLocaleData(resourceManager.getAllResources(new ResourceLocation(s2, s1)));
@@ -120,4 +123,16 @@
          return "Format error: " + s;
       }
    }
+
+   private String injectI18nData(String format, Object... args) {
+      for (Supplier<InputStream> supplier : HyperiumLocale.LANG_FILES.get(args[0].toString())) {
+         try (InputStream stream = supplier.get()) {
+            loadLocaleData(stream);
+         } catch (IOException e) {
+            e.printStackTrace();
+         }
+      }
+
+      return String.format(format, args);
+   }
 }
