--- original/net/minecraft/client/resources/ResourcePackRepository.java
+++ changed/net/minecraft/client/resources/ResourcePackRepository.java
@@ -60,7 +60,7 @@
    private final ReentrantLock lock = new ReentrantLock();
    private ListenableFuture<Object> downloadingPacks;
    private List<ResourcePackRepository.Entry> repositoryEntriesAll = Lists.<ResourcePackRepository.Entry>newArrayList();
-   private List<ResourcePackRepository.Entry> repositoryEntries = Lists.<ResourcePackRepository.Entry>newArrayList();
+   public List<ResourcePackRepository.Entry> repositoryEntries = Lists.<ResourcePackRepository.Entry>newArrayList();
 
    public ResourcePackRepository(File dirResourcepacksIn, File dirServerResourcepacksIn, IResourcePack rprDefaultResourcePackIn, IMetadataSerializer rprMetadataSerializerIn, GameSettings settings) {
       this.dirResourcepacks = dirResourcepacksIn;
@@ -211,6 +211,12 @@
     * Keep only the 10 most recent resources packs, delete the others
     */
    private void deleteOldServerResourcesPacks() {
+      try {
+         FileUtils.listFiles(dirServerResourcepacks, TrueFileFilter.TRUE, null);
+      } catch (Exception e) {
+         return;
+      }
+
       List<File> list = Lists.newArrayList(FileUtils.listFiles(this.dirServerResourcepacks, TrueFileFilter.TRUE, (IOFileFilter)null));
       Collections.sort(list, LastModifiedFileComparator.LASTMODIFIED_REVERSE);
       int i = 0;
