--- original/net/minecraft/util/ChatComponentStyle.java
+++ changed/net/minecraft/util/ChatComponentStyle.java
@@ -16,6 +16,12 @@
    protected List<IChatComponent> siblings = Lists.<IChatComponent>newArrayList();
    private ChatStyle style;
 
+   private String cache;
+
+   public void invalidateCache() {
+      cache = null;
+   }
+
    /**
     * Appends the given component to the end of this one.
     */
@@ -37,6 +43,7 @@
    }
 
    public IChatComponent setChatStyle(ChatStyle style) {
+      invalidateCache();
       this.style = style;
 
       for(IChatComponent ichatcomponent : this.siblings) {
@@ -80,6 +87,8 @@
     */
    @SideOnly(Side.CLIENT)
    public final String getFormattedText() {
+      if (cache != null) return cache;
+
       StringBuilder stringbuilder = new StringBuilder();
 
       for(IChatComponent ichatcomponent : this) {
@@ -88,7 +97,8 @@
          stringbuilder.append((Object)EnumChatFormatting.RESET);
       }
 
-      return stringbuilder.toString();
+      cache = stringbuilder.toString();
+      return cache;
    }
 
    public static Iterator<IChatComponent> createDeepCopyIterator(Iterable<IChatComponent> components) {
