--- original/net/minecraft/entity/Entity.java
+++ changed/net/minecraft/entity/Entity.java
@@ -12,6 +12,7 @@
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.block.state.pattern.BlockPattern;
+import net.minecraft.client.Minecraft;
 import net.minecraft.command.CommandResultStats;
 import net.minecraft.command.ICommandSender;
 import net.minecraft.crash.CrashReport;
@@ -2108,14 +2109,25 @@
       return true;
    }
 
+   private Entity parent;
+   private long nameCacheTime = System.currentTimeMillis();
+   private IChatComponent cachedName;
+
    /**
     * Get the formatted ChatComponent that will be used for the sender's username in chat
     */
    public IChatComponent getDisplayName() {
-      ChatComponentText chatcomponenttext = new ChatComponentText(this.getName());
-      chatcomponenttext.getChatStyle().setChatHoverEvent(this.getHoverEvent());
-      chatcomponenttext.getChatStyle().setInsertion(this.getUniqueID().toString());
-      return chatcomponenttext;
+      if (cachedName == null || System.currentTimeMillis() - nameCacheTime > 50L) {
+         ChatComponentText chatcomponenttext = new ChatComponentText(parent.getName());
+         //not needed otherwise
+         if (Minecraft.getMinecraft().isIntegratedServerRunning())
+            chatcomponenttext.getChatStyle().setChatHoverEvent(getHoverEvent());
+         chatcomponenttext.getChatStyle().setInsertion(parent.getUniqueID().toString());
+         cachedName = chatcomponenttext;
+         nameCacheTime = System.currentTimeMillis();
+      }
+
+      return cachedName;
    }
 
    /**
