--- original/net/minecraft/util/MouseHelper.java
+++ changed/net/minecraft/util/MouseHelper.java
@@ -1,5 +1,6 @@
 package net.minecraft.util;
 
+import cc.hyperium.Hyperium;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.lwjgl.input.Mouse;
@@ -25,7 +26,10 @@
      * Ungrabs the mouse cursor so it can be moved and set it to the center of the screen
      */
     public void ungrabMouseCursor() {
-        Mouse.setCursorPosition(Display.getWidth() / 2, Display.getHeight() / 2);
+        if (Hyperium.INSTANCE.getHandlers().getMouseListener().shouldResetMouse()) {
+            Mouse.setCursorPosition(Display.getWidth() / 2, Display.getHeight() / 2);
+        }
+
         Mouse.setGrabbed(false);
     }
 
