--- original/net/minecraft/client/gui/GuiIngame.java
+++ changed/net/minecraft/client/gui/GuiIngame.java
@@ -6,6 +6,12 @@
 import java.util.Collection;
 import java.util.List;
 import java.util.Random;
+import cc.hyperium.Hyperium;
+import cc.hyperium.config.Settings;
+import cc.hyperium.event.EventBus;
+import cc.hyperium.event.render.RenderHUDEvent;
+import cc.hyperium.event.render.RenderSelectedItemEvent;
+import cc.hyperium.mods.chromahud.displayitems.hyperium.ScoreboardDisplay;
 import net.minecraft.block.material.Material;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.GlStateManager;
@@ -90,6 +96,11 @@
     /** Used with updateCounter to make the heart bar flash */
     private long healthUpdateCounter = 0L;
 
+    public static boolean renderScoreboard = true;
+    public static boolean renderHealth = true;
+    public static boolean renderFood = true;
+    public static boolean renderArmor = true;
+
     public GuiIngame(Minecraft mcIn) {
         this.mc = mcIn;
         this.itemRenderer = mcIn.getRenderItem();
@@ -299,6 +310,11 @@
         GlStateManager.popMatrix();
         scoreobjective1 = scoreboard.getObjectiveInDisplaySlot(0);
 
+        mc.mcProfiler.startSection("hyperium_overlay");
+        EventBus.INSTANCE.post(new RenderHUDEvent(new ScaledResolution(mc), partialTicks));
+        GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+        mc.mcProfiler.endSection();
+
         if (!this.mc.gameSettings.keyBindPlayerList.isKeyDown() || this.mc.isIntegratedServerRunning() && this.mc.thePlayer.sendQueue.getPlayerInfoMap().size() <= 1 && scoreobjective1 == null) {
             this.overlayPlayerList.updatePlayerList(false);
         } else {
@@ -418,6 +434,10 @@
                 GlStateManager.tryBlendFuncSeparate(770, 771, 1, 0);
                 this.getFontRenderer().drawStringWithShadow(s, (float)i, (float)j, 16777215 + (k << 24));
                 GlStateManager.disableBlend();
+
+                EventBus.INSTANCE.post(new RenderSelectedItemEvent(scaledRes));
+                GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+
                 GlStateManager.popMatrix();
             }
         }
@@ -467,286 +487,233 @@
     }
 
     private void renderScoreboard(ScoreObjective objective, ScaledResolution scaledRes) {
-        Scoreboard scoreboard = objective.getScoreboard();
-        Collection<Score> collection = scoreboard.getSortedScores(objective);
-        List<Score> list = Lists.newArrayList(Iterables.filter(collection, new Predicate<Score>() {
-            public boolean apply(Score p_apply_1_) {
-                return p_apply_1_.getPlayerName() != null && !p_apply_1_.getPlayerName().startsWith("#");
-            }
-        }));
-
-        if (list.size() > 15) {
-            collection = Lists.newArrayList(Iterables.skip(list, collection.size() - 15));
-        } else {
-            collection = list;
+        if (!renderScoreboard) {
+            return;
         }
 
-        int i = this.getFontRenderer().getStringWidth(objective.getDisplayName());
-
-        for (Score score : collection) {
-            ScorePlayerTeam scoreplayerteam = scoreboard.getPlayersTeam(score.getPlayerName());
-            String s = ScorePlayerTeam.formatPlayerName(scoreplayerteam, score.getPlayerName()) + ": " + EnumChatFormatting.RED + score.getScorePoints();
-            i = Math.max(i, this.getFontRenderer().getStringWidth(s));
-        }
-
-        int i1 = collection.size() * this.getFontRenderer().FONT_HEIGHT;
-        int j1 = scaledRes.getScaledHeight() / 2 + i1 / 3;
-        int k1 = 3;
-        int l1 = scaledRes.getScaledWidth() - i - k1;
-        int j = 0;
-
-        for (Score score1 : collection) {
-            ++j;
-            ScorePlayerTeam scoreplayerteam1 = scoreboard.getPlayersTeam(score1.getPlayerName());
-            String s1 = ScorePlayerTeam.formatPlayerName(scoreplayerteam1, score1.getPlayerName());
-            String s2 = EnumChatFormatting.RED + "" + score1.getScorePoints();
-            int k = j1 - j * this.getFontRenderer().FONT_HEIGHT;
-            int l = scaledRes.getScaledWidth() - k1 + 2;
-            drawRect(l1 - 2, k, l, k + this.getFontRenderer().FONT_HEIGHT, 1342177280);
-            this.getFontRenderer().drawString(s1, l1, k, 553648127);
-            this.getFontRenderer().drawString(s2, l - this.getFontRenderer().getStringWidth(s2), k, 553648127);
-
-            if (j == collection.size()) {
-                String s3 = objective.getDisplayName();
-                drawRect(l1 - 2, k - this.getFontRenderer().FONT_HEIGHT - 1, l, k - 1, 1610612736);
-                drawRect(l1 - 2, k - 1, l, k, 1342177280);
-                this.getFontRenderer().drawString(s3, l1 + i / 2 - this.getFontRenderer().getStringWidth(s3) / 2, k - this.getFontRenderer().FONT_HEIGHT, 553648127);
-            }
-        }
+        ScoreboardDisplay.objective = objective;
+        ScoreboardDisplay.resolution = scaledRes;
+        Hyperium.INSTANCE.getHandlers().getScoreboardRenderer().render(objective, scaledRes);
     }
 
-    private void renderPlayerStats(ScaledResolution scaledRes) {
-        if (this.mc.getRenderViewEntity() instanceof EntityPlayer) {
-            EntityPlayer entityplayer = (EntityPlayer)this.mc.getRenderViewEntity();
-            int i = MathHelper.ceiling_float_int(entityplayer.getHealth());
-            boolean flag = this.healthUpdateCounter > (long)this.updateCounter && (this.healthUpdateCounter - (long)this.updateCounter) / 3L % 2L == 1L;
+    private void renderPlayerStats(ScaledResolution resolution) {
+        if (mc.getRenderViewEntity() instanceof EntityPlayer) {
+            EntityPlayer entityplayer = (EntityPlayer) mc.getRenderViewEntity();
+            int currentHealth = MathHelper.ceiling_float_int(entityplayer.getHealth());
+            boolean isGettingDamage = healthUpdateCounter > (long) updateCounter && (healthUpdateCounter - (long) updateCounter) / 3L % 2L == 1L;
 
-            if (i < this.playerHealth && entityplayer.hurtResistantTime > 0) {
-                this.lastSystemTime = Minecraft.getSystemTime();
-                this.healthUpdateCounter = (long)(this.updateCounter + 20);
-            } else if (i > this.playerHealth && entityplayer.hurtResistantTime > 0) {
-                this.lastSystemTime = Minecraft.getSystemTime();
-                this.healthUpdateCounter = (long)(this.updateCounter + 10);
+            if (currentHealth < playerHealth && entityplayer.hurtResistantTime > 0) {
+                lastSystemTime = Minecraft.getSystemTime();
+                healthUpdateCounter = updateCounter + 20;
+            } else if (currentHealth > playerHealth && entityplayer.hurtResistantTime > 0) {
+                lastSystemTime = Minecraft.getSystemTime();
+                healthUpdateCounter = updateCounter + 10;
             }
 
-            if (Minecraft.getSystemTime() - this.lastSystemTime > 1000L) {
-                this.playerHealth = i;
-                this.lastPlayerHealth = i;
-                this.lastSystemTime = Minecraft.getSystemTime();
+            if (Minecraft.getSystemTime() - lastSystemTime > 1000L) {
+                playerHealth = currentHealth;
+                lastPlayerHealth = currentHealth;
+                lastSystemTime = Minecraft.getSystemTime();
             }
 
-            this.playerHealth = i;
-            int j = this.lastPlayerHealth;
-            this.rand.setSeed((long)(this.updateCounter * 312871));
-            boolean flag1 = false;
+            playerHealth = currentHealth;
+            int lastPlayerHealth = this.lastPlayerHealth;
+            rand.setSeed(updateCounter * 312871);
+            boolean alwaysFalseFlagWhatIsThis = false;
             FoodStats foodstats = entityplayer.getFoodStats();
-            int k = foodstats.getFoodLevel();
-            int l = foodstats.getPrevFoodLevel();
+            int foodLevel = foodstats.getFoodLevel();
+            int prevFoodLevel = foodstats.getPrevFoodLevel();
             IAttributeInstance iattributeinstance = entityplayer.getEntityAttribute(SharedMonsterAttributes.maxHealth);
-            int i1 = scaledRes.getScaledWidth() / 2 - 91;
-            int j1 = scaledRes.getScaledWidth() / 2 + 91;
-            int k1 = scaledRes.getScaledHeight() - 39;
-            float f = (float)iattributeinstance.getAttributeValue();
-            float f1 = entityplayer.getAbsorptionAmount();
-            int l1 = MathHelper.ceiling_float_int((f + f1) / 2.0F / 10.0F);
-            int i2 = Math.max(10 - (l1 - 2), 3);
-            int j2 = k1 - (l1 - 1) * i2 - 10;
-            float f2 = f1;
-            int k2 = entityplayer.getTotalArmorValue();
-            int l2 = -1;
+            int widthLeft = resolution.getScaledWidth() / 2 - 91;
+            int widthRight = resolution.getScaledWidth() / 2 + 91;
+            int height = resolution.getScaledHeight() - 39;
+            float attributeValue = (float) iattributeinstance.getAttributeValue();
+            float absorptionAmount = entityplayer.getAbsorptionAmount();
+            int extraHealth = MathHelper.ceiling_float_int((attributeValue + absorptionAmount) / 2.0F / 10.0F);
+            int extraHeart = Math.max(10 - (extraHealth - 2), 3);
+            int heartHeight = height - (extraHealth - 1) * extraHeart - 10;
+            float tempAbsorptionAmount = absorptionAmount;
+            int armorValue = entityplayer.getTotalArmorValue();
+            int regeneration = -1;
 
             if (entityplayer.isPotionActive(Potion.regeneration)) {
-                l2 = this.updateCounter % MathHelper.ceiling_float_int(f + 5.0F);
+                regeneration = updateCounter % MathHelper.ceiling_float_int(attributeValue + 5.0F);
             }
 
-            this.mc.mcProfiler.startSection("armor");
+            mc.mcProfiler.startSection("armor");
 
-            for (int i3 = 0; i3 < 10; ++i3) {
-                if (k2 > 0) {
-                    int j3 = i1 + i3 * 8;
+            if (renderArmor) {
+                for (int armorPosition = 0; armorPosition < 10; armorPosition++) {
+                    if (armorValue > 0) {
+                        int j3 = widthLeft + armorPosition * 8;
 
-                    if (i3 * 2 + 1 < k2) {
-                        this.drawTexturedModalRect(j3, j2, 34, 9, 9, 9);
-                    }
+                        if (armorPosition * 2 + 1 < armorValue) {
+                            drawTexturedModalRect(j3, heartHeight, 34, 9, 9, 9);
+                        }
 
-                    if (i3 * 2 + 1 == k2) {
-                        this.drawTexturedModalRect(j3, j2, 25, 9, 9, 9);
-                    }
+                        if (armorPosition * 2 + 1 == armorValue) {
+                            drawTexturedModalRect(j3, heartHeight, 25, 9, 9, 9);
+                        }
 
-                    if (i3 * 2 + 1 > k2) {
-                        this.drawTexturedModalRect(j3, j2, 16, 9, 9, 9);
+                        if (armorPosition * 2 + 1 > armorValue) {
+                            drawTexturedModalRect(j3, heartHeight, 16, 9, 9, 9);
+                        }
                     }
                 }
             }
 
-            this.mc.mcProfiler.endStartSection("health");
+            mc.mcProfiler.endStartSection("health");
 
-            for (int i6 = MathHelper.ceiling_float_int((f + f1) / 2.0F) - 1; i6 >= 0; --i6) {
-                int j6 = 16;
+            for (int healthHeartAmount = MathHelper.ceiling_float_int((attributeValue + absorptionAmount) / 2.0F) - 1; healthHeartAmount >= 0; --healthHeartAmount) {
+                int baseTextureX = 16;
 
                 if (entityplayer.isPotionActive(Potion.poison)) {
-                    j6 += 36;
+                    baseTextureX += 36;
                 } else if (entityplayer.isPotionActive(Potion.wither)) {
-                    j6 += 72;
+                    baseTextureX += 72;
                 }
 
-                int k3 = 0;
+                int gettingDamage = 0;
 
-                if (flag) {
-                    k3 = 1;
-                }
+                if (isGettingDamage) gettingDamage = 1;
 
-                int l3 = MathHelper.ceiling_float_int((float)(i6 + 1) / 10.0F) - 1;
-                int i4 = i1 + i6 % 10 * 8;
-                int j4 = k1 - l3 * i2;
+                int healthInt = MathHelper.ceiling_float_int((float) (healthHeartAmount + 1) / 10.0F) - 1;
+                int healthWidth = widthLeft + healthHeartAmount % 10 * 8;
+                int healthHeight = height - healthInt * extraHeart;
 
-                if (i <= 4) {
-                    j4 += this.rand.nextInt(2);
-                }
+                if (currentHealth <= 4) healthHeight += rand.nextInt(2);
+                if (healthHeartAmount == regeneration) healthHeight -= 2;
+                int hardCore = 0;
 
-                if (i6 == l2) {
-                    j4 -= 2;
-                }
+                if (entityplayer.worldObj.getWorldInfo().isHardcoreModeEnabled()) hardCore = 5;
 
-                int k4 = 0;
+                if (renderHealth) {
+                    drawTexturedModalRect(healthWidth, healthHeight, 16 + gettingDamage * 9, 9 * hardCore, 9, 9);
 
-                if (entityplayer.worldObj.getWorldInfo().isHardcoreModeEnabled()) {
-                    k4 = 5;
-                }
+                    if (!Settings.OLD_HEALTH && isGettingDamage) {
+                        if (healthHeartAmount * 2 + 1 < lastPlayerHealth) {
+                            drawTexturedModalRect(healthWidth, healthHeight, baseTextureX + 54, 9 * hardCore, 9, 9);
+                        }
 
-                this.drawTexturedModalRect(i4, j4, 16 + k3 * 9, 9 * k4, 9, 9);
-
-                if (flag) {
-                    if (i6 * 2 + 1 < j) {
-                        this.drawTexturedModalRect(i4, j4, j6 + 54, 9 * k4, 9, 9);
+                        if (healthHeartAmount * 2 + 1 == lastPlayerHealth) {
+                            drawTexturedModalRect(healthWidth, healthHeight, baseTextureX + 63, 9 * hardCore, 9, 9);
+                        }
                     }
 
-                    if (i6 * 2 + 1 == j) {
-                        this.drawTexturedModalRect(i4, j4, j6 + 63, 9 * k4, 9, 9);
-                    }
-                }
-
-                if (f2 > 0.0F) {
-                    if (f2 == f1 && f1 % 2.0F == 1.0F) {
-                        this.drawTexturedModalRect(i4, j4, j6 + 153, 9 * k4, 9, 9);
+                    if (tempAbsorptionAmount > 0.0F) {
+                        drawTexturedModalRect(healthWidth, healthHeight, tempAbsorptionAmount == absorptionAmount && absorptionAmount % 2.0F
+                                              == 1.0F ? baseTextureX + 153 : baseTextureX + 144, 9 * hardCore, 9, 9);
+                        tempAbsorptionAmount -= 2.0F;
                     } else {
-                        this.drawTexturedModalRect(i4, j4, j6 + 144, 9 * k4, 9, 9);
-                    }
+                        if (healthHeartAmount * 2 + 1 < currentHealth) {
+                            drawTexturedModalRect(healthWidth, healthHeight, baseTextureX + 36, 9 * hardCore, 9, 9);
+                        }
 
-                    f2 -= 2.0F;
-                } else {
-                    if (i6 * 2 + 1 < i) {
-                        this.drawTexturedModalRect(i4, j4, j6 + 36, 9 * k4, 9, 9);
+                        if (healthHeartAmount * 2 + 1 == currentHealth) {
+                            drawTexturedModalRect(healthWidth, healthHeight, baseTextureX + 45, 9 * hardCore, 9, 9);
+                        }
                     }
-
-                    if (i6 * 2 + 1 == i) {
-                        this.drawTexturedModalRect(i4, j4, j6 + 45, 9 * k4, 9, 9);
-                    }
                 }
             }
 
-            Entity entity = entityplayer.ridingEntity;
+            if (renderFood) {
+                Entity entity = entityplayer.ridingEntity;
 
-            if (entity == null) {
-                this.mc.mcProfiler.endStartSection("food");
+                if (entity == null) {
+                    mc.mcProfiler.endStartSection("food");
+                    for (int foodPostion = 0; foodPostion < 10; ++foodPostion) {
+                        int foodHeight = height;
+                        int textureXT = 16;
+                        int textureX = 0;
 
-                for (int k6 = 0; k6 < 10; ++k6) {
-                    int i7 = k1;
-                    int l7 = 16;
-                    int j8 = 0;
+                        if (entityplayer.isPotionActive(Potion.hunger)) {
+                            textureXT += 36;
+                            textureX = 13;
+                        }
 
-                    if (entityplayer.isPotionActive(Potion.hunger)) {
-                        l7 += 36;
-                        j8 = 13;
-                    }
+                        if (entityplayer.getFoodStats().getSaturationLevel() <= 0.0F && updateCounter % (foodLevel * 3 + 1) == 0) {
+                            foodHeight = height + (rand.nextInt(3) - 1);
+                        }
 
-                    if (entityplayer.getFoodStats().getSaturationLevel() <= 0.0F && this.updateCounter % (k * 3 + 1) == 0) {
-                        i7 = k1 + (this.rand.nextInt(3) - 1);
-                    }
+                        if (alwaysFalseFlagWhatIsThis) textureX = 1;
 
-                    if (flag1) {
-                        j8 = 1;
-                    }
+                        int foodPositionX = widthRight - foodPostion * 8 - 9;
+                        drawTexturedModalRect(foodPositionX, foodHeight, 16 + textureX * 9, 27, 9, 9);
 
-                    int i9 = j1 - k6 * 8 - 9;
-                    this.drawTexturedModalRect(i9, i7, 16 + j8 * 9, 27, 9, 9);
+                        if (alwaysFalseFlagWhatIsThis) {
+                            if (foodPostion * 2 + 1 < prevFoodLevel) {
+                                drawTexturedModalRect(foodPositionX, foodHeight, textureXT + 54, 27, 9, 9);
+                            }
 
-                    if (flag1) {
-                        if (k6 * 2 + 1 < l) {
-                            this.drawTexturedModalRect(i9, i7, l7 + 54, 27, 9, 9);
+                            if (foodPostion * 2 + 1 == prevFoodLevel) {
+                                drawTexturedModalRect(foodPositionX, foodHeight, textureXT + 63, 27, 9, 9);
+                            }
                         }
 
-                        if (k6 * 2 + 1 == l) {
-                            this.drawTexturedModalRect(i9, i7, l7 + 63, 27, 9, 9);
+                        if (foodPostion * 2 + 1 < foodLevel) {
+                            drawTexturedModalRect(foodPositionX, foodHeight, textureXT + 36, 27, 9, 9);
                         }
-                    }
 
-                    if (k6 * 2 + 1 < k) {
-                        this.drawTexturedModalRect(i9, i7, l7 + 36, 27, 9, 9);
+                        if (foodPostion * 2 + 1 == foodLevel) {
+                            drawTexturedModalRect(foodPositionX, foodHeight, textureXT + 45, 27, 9, 9);
+                        }
                     }
+                } else if (entity instanceof EntityLivingBase) {
+                    mc.mcProfiler.endStartSection("mountHealth");
+                    EntityLivingBase entitylivingbase = (EntityLivingBase) entity;
+                    int tempHealth = (int) Math.ceil(entitylivingbase.getHealth());
+                    float maxHealth = entitylivingbase.getMaxHealth();
+                    int maxHeart = (int) (maxHealth + 0.5F) / 2;
 
-                    if (k6 * 2 + 1 == k) {
-                        this.drawTexturedModalRect(i9, i7, l7 + 45, 27, 9, 9);
-                    }
-                }
-            } else if (entity instanceof EntityLivingBase) {
-                this.mc.mcProfiler.endStartSection("mountHealth");
-                EntityLivingBase entitylivingbase = (EntityLivingBase)entity;
-                int j7 = (int)Math.ceil((double)entitylivingbase.getHealth());
-                float f3 = entitylivingbase.getMaxHealth();
-                int k8 = (int)(f3 + 0.5F) / 2;
+                    if (maxHeart > 30) maxHeart = 30;
+                    int mountHealthHeight = height;
 
-                if (k8 > 30) {
-                    k8 = 30;
-                }
+                    for (int tempInt = 0; maxHeart > 0; tempInt += 20) {
+                        int heartInt = Math.min(maxHeart, 10);
+                        maxHeart -= heartInt;
 
-                int j9 = k1;
+                        for (int mountHealth = 0; mountHealth < heartInt; ++mountHealth) {
+                            int textureX = 52;
+                            int additionalX = 0;
 
-                for (int k9 = 0; k8 > 0; k9 += 20) {
-                    int l4 = Math.min(k8, 10);
-                    k8 -= l4;
+                            if (alwaysFalseFlagWhatIsThis) {
+                                additionalX = 1;
+                            }
 
-                    for (int i5 = 0; i5 < l4; ++i5) {
-                        int j5 = 52;
-                        int k5 = 0;
+                            int mountHealthPositionX = widthRight - mountHealth * 8 - 9;
+                            drawTexturedModalRect(mountHealthPositionX, mountHealthHeight, textureX + additionalX * 9, 9, 9, 9);
 
-                        if (flag1) {
-                            k5 = 1;
-                        }
+                            if (mountHealth * 2 + 1 + tempInt < tempHealth) {
+                                drawTexturedModalRect(mountHealthPositionX, mountHealthHeight, textureX + 36, 9, 9, 9);
+                            }
 
-                        int l5 = j1 - i5 * 8 - 9;
-                        this.drawTexturedModalRect(l5, j9, j5 + k5 * 9, 9, 9, 9);
-
-                        if (i5 * 2 + 1 + k9 < j7) {
-                            this.drawTexturedModalRect(l5, j9, j5 + 36, 9, 9, 9);
+                            if (mountHealth * 2 + 1 + tempInt == tempHealth) {
+                                drawTexturedModalRect(mountHealthPositionX, mountHealthHeight, textureX + 45, 9, 9, 9);
+                            }
                         }
 
-                        if (i5 * 2 + 1 + k9 == j7) {
-                            this.drawTexturedModalRect(l5, j9, j5 + 45, 9, 9, 9);
-                        }
+                        mountHealthHeight -= 10;
                     }
-
-                    j9 -= 10;
                 }
             }
 
-            this.mc.mcProfiler.endStartSection("air");
+            mc.mcProfiler.endStartSection("air");
 
             if (entityplayer.isInsideOfMaterial(Material.water)) {
-                int l6 = this.mc.thePlayer.getAir();
-                int k7 = MathHelper.ceiling_double_int((double)(l6 - 2) * 10.0D / 300.0D);
-                int i8 = MathHelper.ceiling_double_int((double)l6 * 10.0D / 300.0D) - k7;
-
-                for (int l8 = 0; l8 < k7 + i8; ++l8) {
-                    if (l8 < k7) {
-                        this.drawTexturedModalRect(j1 - l8 * 8 - 9, j2, 16, 18, 9, 9);
+                int air = mc.thePlayer.getAir();
+                int airCheck = MathHelper.ceiling_double_int((double) (air - 2) * 10.0D / 300.0D);
+                int air2 = MathHelper.ceiling_double_int((double) air * 10.0D / 300.0D) - airCheck;
+                int bound = airCheck + air2;
+                for (int airPosition = 0; airPosition < bound; airPosition++) {
+                    if (airPosition < airCheck) {
+                        drawTexturedModalRect(widthRight - airPosition * 8 - 9, heartHeight, 16, 18, 9, 9);
                     } else {
-                        this.drawTexturedModalRect(j1 - l8 * 8 - 9, j2, 25, 18, 9, 9);
+                        drawTexturedModalRect(widthRight - airPosition * 8 - 9, heartHeight, 25, 18, 9, 9);
                     }
                 }
             }
 
-            this.mc.mcProfiler.endSection();
+            mc.mcProfiler.endSection();
         }
     }
 
@@ -754,26 +721,41 @@
      * Renders dragon's (boss) health on the HUD
      */
     private void renderBossHealth() {
-        if (BossStatus.bossName != null && BossStatus.statusBarTime > 0) {
+        if (BossStatus.bossName != null && BossStatus.statusBarTime > 0 && Settings.BOSSBAR_ALL) {
             --BossStatus.statusBarTime;
-            FontRenderer fontrenderer = this.mc.fontRendererObj;
-            ScaledResolution scaledresolution = new ScaledResolution(this.mc);
-            int i = scaledresolution.getScaledWidth();
-            int j = 182;
-            int k = i / 2 - j / 2;
-            int l = (int)(BossStatus.healthScale * (float)(j + 1));
-            int i1 = 12;
-            this.drawTexturedModalRect(k, i1, 0, 74, j, 5);
-            this.drawTexturedModalRect(k, i1, 0, 74, j, 5);
 
-            if (l > 0) {
-                this.drawTexturedModalRect(k, i1, 0, 79, l, 5);
+            FontRenderer fontRenderer = Minecraft.getMinecraft().fontRendererObj;
+            ScaledResolution resolution = new ScaledResolution(Minecraft.getMinecraft());
+            double scaledWidth = resolution.getScaledWidth();
+            double scaledHeight = resolution.getScaledHeight();
+
+            String bossName = BossStatus.bossName;
+
+            if (Settings.BOSSBAR_TEXT) {
+                GlStateManager.pushMatrix();
+                GlStateManager.translate(scaledWidth * Settings.BOSSBAR_X, scaledHeight * Settings.BOSSBAR_Y - (10D * Settings.BOSSBAR_SCALE), 0);
+                GlStateManager.scale(Settings.BOSSBAR_SCALE, Settings.BOSSBAR_SCALE, Settings.BOSSBAR_SCALE);
+                GlStateManager.translate(-fontRenderer.getStringWidth(bossName) / 2F , 0, 0);
+                fontRenderer.drawStringWithShadow(bossName, 0, 0, -1);
+                GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+                Minecraft.getMinecraft().getTextureManager().bindTexture(Gui.icons);
+                GlStateManager.popMatrix();
             }
 
-            String s = BossStatus.bossName;
-            this.getFontRenderer().drawStringWithShadow(s, (float)(i / 2 - this.getFontRenderer().getStringWidth(s) / 2), (float)(i1 - 10), 16777215);
-            GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
-            this.mc.getTextureManager().bindTexture(icons);
+            int widthLocation = 182;
+
+            if (Settings.BOSSBAR_BAR) {
+                int healthScale = (int) (BossStatus.healthScale * (float) (widthLocation + 1) * Settings.BOSSBAR_SCALE);
+
+                GlStateManager.pushMatrix();
+                GlStateManager.translate(Settings.BOSSBAR_X * scaledWidth - widthLocation / 2F * Settings.BOSSBAR_SCALE, Settings.BOSSBAR_Y * scaledHeight , 0);
+                GlStateManager.scale(Settings.BOSSBAR_SCALE, Settings.BOSSBAR_SCALE, Settings.BOSSBAR_SCALE);
+
+                drawTexturedModalRect(0, 0, 0, 74, widthLocation, 5);
+
+                if (healthScale > 0) drawTexturedModalRect(0, 0, 0, 79, widthLocation, 5);
+                GlStateManager.popMatrix();
+            }
         }
     }
 
@@ -948,6 +930,8 @@
     }
 
     public void displayTitle(String title, String subTitle, int timeFadeIn, int displayTime, int timeFadeOut) {
+        if (Settings.HIDE_TITLES) return;
+
         if (title == null && subTitle == null && timeFadeIn < 0 && displayTime < 0 && timeFadeOut < 0) {
             this.displayedTitle = "";
             this.displayedSubTitle = "";
