--- original/net/minecraft/client/gui/GuiControls.java
+++ changed/net/minecraft/client/gui/GuiControls.java
@@ -1,6 +1,10 @@
 package net.minecraft.client.gui;
 
 import java.io.IOException;
+
+import cc.hyperium.Hyperium;
+import cc.hyperium.gui.keybinds.GuiKeybinds;
+import cc.hyperium.handlers.handlers.keybinds.HyperiumBind;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.resources.I18n;
 import net.minecraft.client.settings.GameSettings;
@@ -10,7 +14,7 @@
 
 @SideOnly(Side.CLIENT)
 public class GuiControls extends GuiScreen {
-    private static final GameSettings.Options[] optionsArr = new GameSettings.Options[] {GameSettings.Options.INVERT_MOUSE, GameSettings.Options.SENSITIVITY, GameSettings.Options.TOUCHSCREEN};
+    private static final GameSettings.Options[] optionsArr = new GameSettings.Options[]{GameSettings.Options.INVERT_MOUSE, GameSettings.Options.SENSITIVITY, GameSettings.Options.TOUCHSCREEN};
     /** A reference to the screen object that created this. Used for navigating between screens. */
     private GuiScreen parentScreen;
     protected String screenTitle = "Controls";
@@ -61,6 +65,7 @@
      * Called by the controls from the buttonList when activated. (Mouse pressed for buttons)
      */
     protected void actionPerformed(GuiButton button) throws IOException {
+        System.out.println(button.id);
         if (button.id == 200) {
             this.mc.displayGuiScreen(this.parentScreen);
         } else if (button.id == 201) {
@@ -70,7 +75,7 @@
 
             KeyBinding.resetKeyBindingArrayAndHash();
         } else if (button.id < 100 && button instanceof GuiOptionButton) {
-            this.options.setOptionValue(((GuiOptionButton)button).returnEnumOptions(), 1);
+            this.options.setOptionValue(((GuiOptionButton) button).returnEnumOptions(), 1);
             button.displayString = this.options.getKeyBinding(GameSettings.Options.getEnumOptions(button.id));
         }
     }
@@ -106,6 +111,9 @@
             if (keyCode == 1) {
                 this.options.setOptionKeyBinding(this.buttonId, 0);
             } else if (keyCode != 0) {
+                if (buttonId.getKeyCategory().contains("hyperium"))
+                    Hyperium.INSTANCE.getHandlers().getKeybindHandler().getBinding(buttonId.getKeyDescription()).setKeyCode(keyCode);
+
                 this.options.setOptionKeyBinding(this.buttonId, keyCode);
             } else if (typedChar > 0) {
                 this.options.setOptionKeyBinding(this.buttonId, typedChar + 256);
