--- original/net/minecraft/client/gui/GuiScreenBook.java
+++ changed/net/minecraft/client/gui/GuiScreenBook.java
@@ -541,4 +541,20 @@
                 }
             }
         }
+
+    public NBTTagList getBookPages() {
+        return bookPages;
+    }
+
+    public void setBookPages(NBTTagList bookPages) {
+        this.bookPages = bookPages;
+    }
+
+    public int getCurrPage() {
+        return currPage;
+    }
+
+    public void setCurrPage(int currPage) {
+        this.currPage = currPage;
+    }
 }
