--- original/net/minecraft/scoreboard/Scoreboard.java
+++ changed/net/minecraft/scoreboard/Scoreboard.java
@@ -236,7 +236,8 @@
      * Removes the team from the scoreboard, updates all player memberships and broadcasts the deletion to all players
      */
     public void removeTeam(ScorePlayerTeam p_96511_1_) {
-        this.teams.remove(p_96511_1_.getRegisteredName());
+        if (p_96511_1_ == null) return;
+        if (p_96511_1_.getRegisteredName() != null) teams.remove(p_96511_1_.getRegisteredName());
 
         for (String s : p_96511_1_.getMembershipCollection()) {
             this.teamMemberships.remove(s);
