--- original/net/minecraft/client/renderer/entity/layers/LayerHeldItem.java
+++ changed/net/minecraft/client/renderer/entity/layers/LayerHeldItem.java
@@ -1,5 +1,6 @@
 package net.minecraft.client.renderer.entity.layers;
 
+import cc.hyperium.mods.oldanimations.OldBlocking;
 import net.minecraft.block.Block;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.model.ModelBiped;
@@ -18,49 +19,14 @@
 @SideOnly(Side.CLIENT)
 public class LayerHeldItem implements LayerRenderer<EntityLivingBase> {
     private final RendererLivingEntity<?> livingEntityRenderer;
+    private OldBlocking oldBlocking = new OldBlocking();
 
     public LayerHeldItem(RendererLivingEntity<?> livingEntityRendererIn) {
         this.livingEntityRenderer = livingEntityRendererIn;
     }
 
     public void doRenderLayer(EntityLivingBase entitylivingbaseIn, float p_177141_2_, float p_177141_3_, float partialTicks, float p_177141_5_, float p_177141_6_, float p_177141_7_, float scale) {
-        ItemStack itemstack = entitylivingbaseIn.getHeldItem();
-
-        if (itemstack != null) {
-            GlStateManager.pushMatrix();
-
-            if (this.livingEntityRenderer.getMainModel().isChild) {
-                float f = 0.5F;
-                GlStateManager.translate(0.0F, 0.625F, 0.0F);
-                GlStateManager.rotate(-20.0F, -1.0F, 0.0F, 0.0F);
-                GlStateManager.scale(f, f, f);
-            }
-
-            ((ModelBiped)this.livingEntityRenderer.getMainModel()).postRenderArm(0.0625F);
-            GlStateManager.translate(-0.0625F, 0.4375F, 0.0625F);
-
-            if (entitylivingbaseIn instanceof EntityPlayer && ((EntityPlayer)entitylivingbaseIn).fishEntity != null) {
-                itemstack = new ItemStack(Items.fishing_rod, 0);
-            }
-
-            Item item = itemstack.getItem();
-            Minecraft minecraft = Minecraft.getMinecraft();
-
-            if (item instanceof ItemBlock && Block.getBlockFromItem(item).getRenderType() == 2) {
-                GlStateManager.translate(0.0F, 0.1875F, -0.3125F);
-                GlStateManager.rotate(20.0F, 1.0F, 0.0F, 0.0F);
-                GlStateManager.rotate(45.0F, 0.0F, 1.0F, 0.0F);
-                float f1 = 0.375F;
-                GlStateManager.scale(-f1, -f1, f1);
-            }
-
-            if (entitylivingbaseIn.isSneaking()) {
-                GlStateManager.translate(0.0F, 0.203125F, 0.0F);
-            }
-
-            minecraft.getItemRenderer().renderItem(entitylivingbaseIn, itemstack, ItemCameraTransforms.TransformType.THIRD_PERSON);
-            GlStateManager.popMatrix();
-        }
+        oldBlocking.doRenderLayer(entitylivingbaseIn, livingEntityRenderer);
     }
 
     public boolean shouldCombineTextures() {
