--- original/net/minecraft/client/gui/GuiScreenOptionsSounds.java
+++ changed/net/minecraft/client/gui/GuiScreenOptionsSounds.java
@@ -1,5 +1,6 @@
 package net.minecraft.client.gui;
 
+import java.awt.*;
 import java.io.IOException;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.audio.PositionedSoundRecord;
@@ -110,8 +111,8 @@
                 }
 
                 GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
-                this.drawTexturedModalRect(this.xPosition + (int)(this.field_146156_o * (float)(this.width - 8)), this.yPosition, 0, 66, 4, 20);
-                this.drawTexturedModalRect(this.xPosition + (int)(this.field_146156_o * (float)(this.width - 8)) + 4, this.yPosition, 196, 66, 4, 20);
+                drawRect(xPosition + (int) (field_146156_o * (float) (width - 6)), yPosition,
+                         xPosition + (int) (field_146156_o * (float) (width - 6)) + 6, yPosition + 20, new Color(0, 0, 0, 150).getRGB());
             }
         }
 
