--- original/net/minecraft/client/renderer/GlStateManager.java
+++ changed/net/minecraft/client/renderer/GlStateManager.java
@@ -284,6 +284,10 @@
       return GL11.glGenTextures();
    }
 
+   public static boolean isTexture2DEnabled() {
+      return textureState[activeTextureUnit].texture2DState.currentState;
+   }
+
    public static void deleteTexture(int texture) {
       GL11.glDeleteTextures(texture);
 
@@ -478,7 +482,7 @@
    }
 
    @SideOnly(Side.CLIENT)
-   static class BooleanState {
+   public static class BooleanState {
       private final int capability;
       private boolean currentState = false;
 
@@ -699,7 +703,7 @@
    }
 
    @SideOnly(Side.CLIENT)
-   static class TextureState {
+   public static class TextureState {
       public GlStateManager.BooleanState texture2DState;
       public int textureName;
 
