--- original/net/minecraft/client/renderer/entity/layers/LayerArmorBase.java
+++ changed/net/minecraft/client/renderer/entity/layers/LayerArmorBase.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.renderer.entity.layers;
 
+import cc.hyperium.config.Settings;
+import cc.hyperium.utils.ItemUtil;
 import com.google.common.collect.Maps;
 import java.util.Map;
 import net.minecraft.client.model.ModelBase;
@@ -38,13 +40,19 @@
     }
 
     public boolean shouldCombineTextures() {
-        return false;
+        return Settings.OLD_ARMOUR;
     }
 
     private void renderLayer(EntityLivingBase entitylivingbaseIn, float p_177182_2_, float p_177182_3_, float partialTicks, float p_177182_5_, float p_177182_6_, float p_177182_7_, float scale, int armorSlot) {
         ItemStack itemstack = this.getCurrentArmor(entitylivingbaseIn, armorSlot);
 
         if (itemstack != null && itemstack.getItem() instanceof ItemArmor) {
+            if (Settings.HIDE_LEATHER_ARMOR && ItemUtil.INSTANCE.getLeatherArmorCollection().contains(itemstack.getItem())) {
+                return;
+            }
+        }
+
+        if (itemstack != null && itemstack.getItem() instanceof ItemArmor) {
             ItemArmor itemarmor = (ItemArmor)itemstack.getItem();
             T t = this.getArmorModel(armorSlot);
             t.setModelAttributes(this.renderer.getMainModel());
@@ -89,6 +97,8 @@
     }
 
     private void renderGlint(EntityLivingBase entitylivingbaseIn, T modelbaseIn, float p_177183_3_, float p_177183_4_, float partialTicks, float p_177183_6_, float p_177183_7_, float p_177183_8_, float scale) {
+        if (Settings.DISABLE_ENCHANT_GLINT) return;
+
         float f = (float)entitylivingbaseIn.ticksExisted + partialTicks;
         this.renderer.bindTexture(ENCHANTED_ITEM_GLINT_RES);
         GlStateManager.enableBlend();
