--- original/net/minecraft/network/play/client/C01PacketChatMessage.java
+++ changed/net/minecraft/network/play/client/C01PacketChatMessage.java
@@ -1,6 +1,7 @@
 package net.minecraft.network.play.client;
 
 import java.io.IOException;
+import net.minecraft.client.gui.GuiChat;
 import net.minecraft.network.Packet;
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.INetHandlerPlayServer;
@@ -12,8 +13,8 @@
     }
 
     public C01PacketChatMessage(String messageIn) {
-        if (messageIn.length() > 100) {
-            messageIn = messageIn.substring(0, 100);
+        if (messageIn.length() > GuiChat.maxStringLength) {
+            messageIn = messageIn.substring(0, GuiChat.maxStringLength);
         }
 
         this.message = messageIn;
@@ -23,7 +24,7 @@
      * Reads the raw packet data from the data stream.
      */
     public void readPacketData(PacketBuffer buf) throws IOException {
-        this.message = buf.readStringFromBuffer(100);
+        this.message = buf.readStringFromBuffer(GuiChat.maxStringLength);
     }
 
     /**
@@ -43,4 +44,8 @@
     public String getMessage() {
         return this.message;
     }
+
+    public void setMessage(String message) {
+        this.message = message;
+    }
 }
