--- original/net/minecraft/server/dedicated/DedicatedServer.java
+++ changed/net/minecraft/server/dedicated/DedicatedServer.java
@@ -30,6 +30,7 @@
 import net.minecraft.server.management.PreYggdrasilConverter;
 import net.minecraft.util.BlockPos;
 import net.minecraft.util.CryptManager;
+import net.minecraft.util.IChatComponent;
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.EnumDifficulty;
 import net.minecraft.world.World;
@@ -570,4 +571,9 @@
       this.commandManager.executeCommand(RConConsoleSource.getInstance(), command);
       return RConConsoleSource.getInstance().getLogContents();
    }
+
+   @Override
+   public void addChatMessage(IChatComponent component) {
+      logger.info(component.getFormattedText());
+   }
 }
