--- original/net/minecraft/client/gui/GuiScreenServerList.java
+++ changed/net/minecraft/client/gui/GuiScreenServerList.java
@@ -1,6 +1,8 @@
 package net.minecraft.client.gui;
 
 import java.io.IOException;
+
+import cc.hyperium.Hyperium;
 import net.minecraft.client.multiplayer.ServerData;
 import net.minecraft.client.resources.I18n;
 import net.minecraftforge.fml.relauncher.Side;
@@ -58,6 +60,7 @@
             if (button.id == 1) {
                 this.field_146303_a.confirmClicked(false, 0);
             } else if (button.id == 0) {
+                Hyperium.INSTANCE.setFromMainMenu(false); // Hyperium
                 this.field_146301_f.serverIP = this.field_146302_g.getText();
                 this.field_146303_a.confirmClicked(true, 0);
             }
