--- original/net/minecraft/client/gui/inventory/GuiInventory.java
+++ changed/net/minecraft/client/gui/inventory/GuiInventory.java
@@ -1,6 +1,9 @@
 package net.minecraft.client.gui.inventory;
 
 import java.io.IOException;
+
+import cc.hyperium.event.EventBus;
+import cc.hyperium.event.interact.ActionPerformedEvent;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.GuiButton;
 import net.minecraft.client.gui.achievement.GuiAchievements;
@@ -127,6 +130,7 @@
      * Called by the controls from the buttonList when activated. (Mouse pressed for buttons)
      */
     protected void actionPerformed(GuiButton button) throws IOException {
+        EventBus.INSTANCE.post(new ActionPerformedEvent(this, button)); // Hyperium
         if (button.id == 0) {
             this.mc.displayGuiScreen(new GuiAchievements(this, this.mc.thePlayer.getStatFileWriter()));
         }
