--- original/net/minecraft/client/gui/GuiOverlayDebug.java
+++ changed/net/minecraft/client/gui/GuiOverlayDebug.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.gui;
 
+import cc.hyperium.config.Settings;
+import cc.hyperium.utils.DebugOverlayUtil;
 import com.google.common.base.Strings;
 import com.google.common.collect.Lists;
 import java.util.List;
@@ -44,6 +46,15 @@
    public void renderDebugInfo(ScaledResolution scaledResolutionIn) {
       this.mc.mcProfiler.startSection("debug");
       GlStateManager.pushMatrix();
+
+      if (Settings.OLD_DEBUG) {
+         DebugOverlayUtil.INSTANCE.renderOldDebugInfoLeft();
+         DebugOverlayUtil.INSTANCE.renderOldDebugInfoRight(scaledResolutionIn);
+         GlStateManager.popMatrix();
+         mc.mcProfiler.endSection();
+         return;
+      }
+
       this.renderDebugInfoLeft();
       this.renderDebugInfoRight(scaledResolutionIn);
       GlStateManager.popMatrix();
