--- original/net/minecraft/nbt/NBTTagCompound.java
+++ changed/net/minecraft/nbt/NBTTagCompound.java
@@ -63,6 +63,7 @@
      * Stores the given tag into the map with the given string key. This is mostly used to store tag lists.
      */
     public void setTag(String key, NBTBase value) {
+        if (value == null) throw new IllegalArgumentException("Invalid null NBT value with key " + key);
         this.tagMap.put(key, value);
     }
 
