--- original/net/minecraft/network/play/server/S2EPacketCloseWindow.java
+++ changed/net/minecraft/network/play/server/S2EPacketCloseWindow.java
@@ -1,6 +1,9 @@
 package net.minecraft.network.play.server;
 
 import java.io.IOException;
+
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.gui.GuiChat;
 import net.minecraft.network.Packet;
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.INetHandlerPlayClient;
@@ -19,6 +22,7 @@
      * Passes this Packet on to the NetHandler for processing.
      */
     public void processPacket(INetHandlerPlayClient handler) {
+        if (Minecraft.getMinecraft().currentScreen instanceof GuiChat) return;
         handler.handleCloseWindow(this);
     }
 
