--- original/net/minecraft/client/resources/DefaultPlayerSkin.java
+++ changed/net/minecraft/client/resources/DefaultPlayerSkin.java
@@ -46,6 +46,6 @@
     * Checks if a players skin model is slim or the default. The Alex model is slime while the Steve model is default.
     */
    private static boolean isSlimSkin(UUID playerUUID) {
-      return (playerUUID.hashCode() & 1) == 1;
+      return playerUUID != null && (playerUUID.hashCode() & 1) == 1;
    }
 }
