--- original/net/minecraft/client/multiplayer/ChunkProviderClient.java
+++ changed/net/minecraft/client/multiplayer/ChunkProviderClient.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.multiplayer;
 
+import cc.hyperium.event.EventBus;
+import cc.hyperium.event.world.chunk.ChunkLoadEvent;
 import com.google.common.collect.Lists;
 import java.util.List;
 import net.minecraft.entity.EnumCreatureType;
@@ -66,6 +68,10 @@
       Chunk chunk = new Chunk(this.worldObj, chunkX, chunkZ);
       this.chunkMapping.add(ChunkCoordIntPair.chunkXZ2Int(chunkX, chunkZ), chunk);
       this.chunkListing.add(chunk);
+
+      // Hyperium - Post chunk loading event
+      EventBus.INSTANCE.post(new ChunkLoadEvent(chunk));
+
       chunk.setChunkLoaded(true);
       return chunk;
    }
