--- original/net/minecraft/block/BlockBeacon.java
+++ changed/net/minecraft/block/BlockBeacon.java
@@ -101,8 +101,8 @@
    }
 
    public static void updateColorAsync(final World worldIn, final BlockPos glassPos) {
-      HttpUtil.field_180193_a.submit(new Runnable() {
-         public void run() {
+/*      HttpUtil.field_180193_a.submit(new Runnable() {
+         public void run() {*/
             Chunk chunk = worldIn.getChunkFromBlockCoords(glassPos);
 
             for(int i = glassPos.getY() - 1; i >= 0; --i) {
@@ -111,7 +111,7 @@
                   break;
                }
 
-               IBlockState iblockstate = worldIn.getBlockState(blockpos);
+               IBlockState iblockstate = chunk.getBlockState(blockpos);
                if(iblockstate.getBlock() == Blocks.beacon) {
                   ((WorldServer)worldIn).addScheduledTask(new Runnable() {
                      public void run() {
@@ -127,6 +127,6 @@
             }
 
          }
-      });
-   }
+//      });
+//   }
 }
