--- original/net/minecraft/client/renderer/entity/RenderLightningBolt.java
+++ changed/net/minecraft/client/renderer/entity/RenderLightningBolt.java
@@ -1,6 +1,8 @@
 package net.minecraft.client.renderer.entity;
 
 import java.util.Random;
+
+import cc.hyperium.config.Settings;
 import net.minecraft.client.renderer.GlStateManager;
 import net.minecraft.client.renderer.Tessellator;
 import net.minecraft.client.renderer.WorldRenderer;
@@ -22,6 +24,8 @@
     * Renders the desired {@code T} type Entity.
     */
    public void doRender(EntityLightningBolt entity, double x, double y, double z, float entityYaw, float partialTicks) {
+      if (Settings.DISABLE_LIGHTNING) return;
+
       Tessellator tessellator = Tessellator.getInstance();
       WorldRenderer worldrenderer = tessellator.getWorldRenderer();
       GlStateManager.disableTexture2D();
