--- original/net/minecraft/client/renderer/tileentity/TileEntityEndPortalRenderer.java
+++ changed/net/minecraft/client/renderer/tileentity/TileEntityEndPortalRenderer.java
@@ -2,6 +2,8 @@
 
 import java.nio.FloatBuffer;
 import java.util.Random;
+
+import cc.hyperium.config.Settings;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.ActiveRenderInfo;
 import net.minecraft.client.renderer.GLAllocation;
@@ -23,6 +25,8 @@
    FloatBuffer field_147528_b = GLAllocation.createDirectFloatBuffer(16);
 
    public void renderTileEntityAt(TileEntityEndPortal te, double x, double y, double z, float partialTicks, int destroyStage) {
+      if (Settings.DISABLE_END_PORTALS) return;
+
       float f = (float)this.rendererDispatcher.entityX;
       float f1 = (float)this.rendererDispatcher.entityY;
       float f2 = (float)this.rendererDispatcher.entityZ;
