--- original/net/minecraft/client/gui/inventory/GuiEditSign.java
+++ changed/net/minecraft/client/gui/inventory/GuiEditSign.java
@@ -37,6 +37,7 @@
     * window resizes, the buttonList is cleared beforehand.
     */
    public void initGui() {
+       super.initGui();
       this.buttonList.clear();
       Keyboard.enableRepeatEvents(true);
       this.buttonList.add(this.doneBtn = new GuiButton(0, this.width / 2 - 100, this.height / 4 + 120, I18n.format("gui.done", new Object[0])));
@@ -47,6 +48,7 @@
     * Called when the screen is unloaded. Used to disable keyboard repeat events
     */
    public void onGuiClosed() {
+       super.onGuiClosed();
       Keyboard.enableRepeatEvents(false);
       NetHandlerPlayClient nethandlerplayclient = this.mc.getNetHandler();
       if(nethandlerplayclient != null) {
